server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -3883,6 +3893,14 @@ private Tuple<BlobStoreIndexShardSnapshots, ShardGeneration> buildBlobStoreIndex
                     generation
                 );
             } catch (NoSuchFileException noSuchFileException) {
+                final long latestGeneration = latestIndexBlobId();
+                final RepositoryData currentRepositoryData = getRepositoryData(latestGeneration);
+                if (currentRepositoryData.shardGenerations().hasShardGen(new RepositoryShardId(indexId, shardId)) == false) {
+                    // Master has concurrently mutated the shard generation. This can happen when master fails over
```

Comment:
I think we're only checking whether the shard has been totally removed here? If the shard still exists then `hasShardGen()` should return `true` but `getShardGen()` will return a different generation from the one we were trying to load.

Reply:
You are right. I used the wrong method. Got confused by its name. It's now obvious that it does not take a shardGen parameter 🤦 
Pushed fe1cf86524225038ef1b9ea31ce0c2b227223c87
Thanks for spotting it.